### DISCUSSION TITLE
Sending email doesn't work

### DISCUSSION THREAD

**Discussion by vicman1200** [elsa-workflows/elsa-core#3353] [Q&A]
Hello Team!

I've had some frustrations, I can't get Elsa to send emails.

I've configured appsettings.json as the documentation says, I even downloaded "ElsaGuides.ContentApproval.Web" and still can't get it to send email.

The SMTP data is correct because I did a test in DocumentApprovalWorkflow.cs with a simple code to send emails with SmtpClient and it sends it from there.

Does anyone have any suggestion?
Has anyone had the same thing happen?

Thanks in advance.

**Comment by sfmskywalker** [comment]
Sorry to hear about frustrations!

Can you provide a little bit more details as to what you mean exactly when you say: can't get Elsa to send emails?
Are you getting any errors? Or nothing happens whatsoever? Is there anything you can observe

Furthermore, steps to reproduce would be mighty helpful to help understand what may be the issue. For example, can you reproduce the issue with a local SMTP host such as smtp4dev? If yes, then the easiest thing would be for you to provide a complete sample project where the send email stuff doesn't work. Then someone can try it on their local machine and troubleshoot.

Most likely, something isn't right with your SMTP configuration. If you can share that, maybe it becomes instantly obvious what the issue might be.

Note also that Elsa uses the MailKit library ad requires you to set the `RequireCredentials` options to `true` if your SMTP host requires authentication (and if so, the UserName and Password values).

**Reply by vicman1200** [reply]
Thank you for your swift response

No error is thrown, the Workflow looks good, I'm following this guide: **ElsaGuides.ContentApproval.Web**. when I post the message in Postman to: _/v1/documents_:

`{
    "Id": "3",
    "Author": {
        "Name": "Felipe Villanueva",
        "Email": "john@gmail.com"
    },
    "Body": "This is sample document."
}`

I receive:

```
<h1>Request for Approval Sent</h1>
<p>Your document has been received and will be reviewed shortly.</p>

```

But the mail is not sent.

Next is what I have configured


**Startup.cs**


```
 public void ConfigureServices(IServiceCollection services)
        {
            //.AddEmailActivities(elsaSection.GetSection("Smtp").Bind)

          //  var elsaSection = Configuration.GetSection("Elsa");
            services
              .AddElsa(elsa => elsa
                  .UseEntityFrameworkPersistence(ef => ef.UseSqlite())
                  .AddConsoleActivities()
                  .AddHttpActivities(opt => { opt.BaseUrl = new Uri("https://localhost:5001"); })
                  .AddEmailActivities(options =>
                        {
                            options.Host = "smtp.sendgrid.net";
                            options.Port = 587;
                            options.UserName = "apiKey";
                            options.Password = "SG.*****";
                            options.RequireCredentials = true;
                            options.UseDefaultCredentials = false;
                            options.SecureSocketOptions = MailKit.Security.SecureSocketOptions.StartTls;
                            
                        })
                  .AddQuartzTemporalActivities()
                  .AddWorkflowsFrom<Startup>()
              );

            services.AddElsaApiEndpoints();
            services.AddRazorPages();
        }
```

Thank you 👋

**Reply by johnwc** [reply]
Have you looked at your sendgrid logs to see if the connection is even being attempted?

**Reply by vicman1200** [reply]
Hi @johnwc 👋

Yes, and there are no error logs or anything, just the attempts I made using the SmtpClient which are all successful

It seems as if the message did not go out

**Reply by johnwc** [reply]
I think you need to set `opt.DeliveryMethod = Elsa.Activities.Email.Options.SmtpDeliveryMethod.Network;` else it may try to use the file system delivery method.

**Reply by johnwc** [reply]
@vicman1200 did that resolve your issue?

**Reply by vicman1200** [reply]
Sorry for my late response

Unfortunately not.

I think I tried all the possibilities.

**Reply by SeeOnTime** [reply]
opt.BaseUrl = new Uri("https://localhost:5001");   try it    opt.BaseUrl = new Uri("http://localhost:5001");

**Reply by johnwc** [reply]
@SeeOnTime I think you may have posted in the wrong discussion thread, as there is no `BaseUrl` property for the [SmtpOptions](https://github.com/elsa-workflows/elsa-core/blob/master/src/activities/Elsa.Activities.Email/Options/SmtpOptions.cs#L5) object.

**Reply by SeeOnTime** [reply]
@johnwc  Thank you for your reminder. I made a mistake. This is the problem I encountered earlier in this example. When the program starts, it's Http, and I set Https, which causes signal_ url Approve click for URL binding is invalid.

**Comment by johnwc** [comment]
You also need to make sure that you are setting things up in the startup, not just setting values in the appsettings.json.

```csharp
.AddElsa(elsa => elsa
    .AddEmailActivities(opt =>
    {
        opt.DeliveryMethod = Elsa.Activities.Email.Options.SmtpDeliveryMethod.Network;
        opt.Host = cfgElsa.Mail!.Host;
        opt.Port = cfgElsa.Mail!.Port ?? 25;
        ...
    })
    ...
```
